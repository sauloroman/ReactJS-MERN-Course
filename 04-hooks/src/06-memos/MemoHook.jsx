import { useMemo, useState } from "react"
import { useCounter } from "../hooks"

const heavyStuff = ( iterationTimes = 1000 ) => {

  for( let i = 1; i <= iterationTimes; i++ ) {
    console.log('Ahi vamos')
  }

  return `${iterationTimes} iteraciones hechas`
}

export const MemoHook = () => {
  
  const [isShown, setIsShown] = useState(false)
  const { counter, onIncrementCounter } = useCounter(4000)
  const memorizedValue = useMemo(() => heavyStuff(counter), [counter])

  return (
    <>
      <h2 className="heading">UseMemo</h2>

      <p className="counter__value">Counter: {counter}</p>

      <button className="btn" onClick={ () => onIncrementCounter() }>+1</button>

      <button
        className="btn"
        onClick={ () => setIsShown( !isShown )}
      >Show | Hide {JSON.stringify(isShown)}</button>

      <span>{memorizedValue}</span>
    </>
  )
}
