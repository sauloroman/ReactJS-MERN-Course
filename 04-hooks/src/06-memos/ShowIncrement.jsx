import React from "react"

export const ShowIncrement = React.memo(({ incrementFn }) => {
  console.log('Render ShowIncrement')
  return (
    <button className="btn" onClick={ () => incrementFn(2) }>Incrementar contador</button>
  )
})
