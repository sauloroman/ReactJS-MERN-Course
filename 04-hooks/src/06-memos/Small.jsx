import React from 'react'

export const Small = React.memo(({ value }) => {
  console.log('Render Small')
  return (
    <div className="counter">
      <p className="counter__value">{value}</p>
      <div className="counter__name">Contador</div>
    </div>
  )
})
