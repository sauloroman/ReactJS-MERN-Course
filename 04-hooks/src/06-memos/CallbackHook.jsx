import { useCallback, useState } from "react"
import { ShowIncrement } from "./ShowIncrement"

export const CallbackHook = () => {

  const [counter, setCounter] = useState(10)

  const increment = useCallback(( value ) => setCounter( counter => counter + value ), []);

  // const increment = ( value ) => {
  //   setCounter( counter => counter + value )
  // }

  return (
    <>
      <h1 className="heading">UseCallback</h1>

      <p className="counter">Counter: {counter}</p>

      <ShowIncrement incrementFn={ increment } />
    </>
  )
}
