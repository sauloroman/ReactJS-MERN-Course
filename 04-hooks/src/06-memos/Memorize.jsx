import { useState } from 'react'
import { useCounter } from '../hooks'
import { Small } from './Small'

export const Memorize = () => {
  
  const { counter, onIncrementCounter } = useCounter(10)
  const [isShown, setIsShown] = useState(false)

  return (
    <>
      <h1 className="heading">Memo Function</h1>

      <Small value={counter} />

      <button className="btn" onClick={() => onIncrementCounter()}>+1</button>

      <button onClick={() => setIsShown(!isShown)} className='btn'>Shown | Hidden {JSON.stringify( isShown )}</button>
    </>
  )
}
