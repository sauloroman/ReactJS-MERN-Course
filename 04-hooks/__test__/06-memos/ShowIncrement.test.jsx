import { fireEvent, render, screen } from '@testing-library/react'
import { ShowIncrement } from '../../src/06-memos/ShowIncrement'

describe('Tests in ShowIncrement', () => {

  const incrementFnMock = jest.fn()

  test('It should match with snapshot', () => {
    const { container } = render(<ShowIncrement incrementFn={incrementFnMock} />)
    expect( container ).toMatchSnapshot()
  })

  test('it should called increment function with an argument of 2', () => {

    render(<ShowIncrement incrementFn={incrementFnMock} />)

    const button = screen.getByText('Incrementar contador')
    fireEvent.click( button )

    expect( incrementFnMock ).toHaveBeenCalledWith(2)

  })

})