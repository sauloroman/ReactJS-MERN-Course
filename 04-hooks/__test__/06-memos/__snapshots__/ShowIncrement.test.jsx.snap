// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in ShowIncrement It should match with snapshot 1`] = `
<div>
  <button
    class="btn"
  >
    Incrementar contador
  </button>
</div>
`;
